### PR TITLE
Enhance multiclass_nms op to support LoD for dygraph mode

### PR DESCRIPTION
**Pull request by yghstill**
### PR types

Function optimization
### PR changes

OPs
### Describe

Enhance multiclass_nms op to support LoD for dygraph mode

### CODE REVIEW

File: paddle/fluid/operators/detection/multiclass_nms_op.cc [hunk 1]
```diff
@@ -21,6 +21,21 @@ namespace operators {
 using Tensor = framework::Tensor;
 using LoDTensor = framework::LoDTensor;
 
+inline std::vector<size_t> GetLodFromRoisNum(const Tensor* rois_num) {
+  std::vector<size_t> rois_lod;
+  auto* rois_num_data = rois_num->data<int>();
+  Tensor cpu_tensor;
+  if (platform::is_gpu_place(rois_num->place())) {
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/multiclass_nms_op.cc]: the rois_num is only on cpu place.

File: paddle/fluid/operators/detection/multiclass_nms_op.cc [hunk 2]
```diff
@@ -535,6 +576,41 @@ class MultiClassNMS2OpMaker : public MultiClassNMSOpMaker {
...
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    MultiClassNMS2Op::InferShape(ctx);
+
+    auto box_dims = ctx->GetInputDim("BBoxes");
+    auto score_dims = ctx->GetInputDim("Scores");
+    auto score_size = score_dims.size();
+    if (score_size == 3) {
+      ctx->SetOutputDim("NmsRoisNum", {box_dims[1], 1});
+    } else {
+      ctx->SetOutputDim("NmsRoisNum", {-1, 1});
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/multiclass_nms_op.cc]: The rank should be 1 and the shape is {-1}

Comment (yghstill) [reply paddle/fluid/operators/detection/multiclass_nms_op.cc]: Done, thanks.

File: paddle/fluid/operators/detection/multiclass_nms_op.cc (hunk 1 above)
Comment (yghstill) [reply paddle/fluid/operators/detection/multiclass_nms_op.cc]: Done, thanks.

**Review by zhiqiu — DISMISSED** [review]
LGTM for op_function_generator.cc

File: paddle/fluid/operators/detection/multiclass_nms_op.cc [hunk 3]
```diff
@@ -535,6 +571,41 @@ class MultiClassNMS2OpMaker : public MultiClassNMSOpMaker {
...
+      : MultiClassNMS2Op(type, inputs, outputs, attrs) {}
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    MultiClassNMS2Op::InferShape(ctx);
+
+    auto box_dims = ctx->GetInputDim("BBoxes");
+    auto score_dims = ctx->GetInputDim("Scores");
+    auto score_size = score_dims.size();
+    if (score_size == 3) {
+      ctx->SetOutputDim("NmsRoisNum", {box_dims[1], 1});
```
Comment (jerrywgz) [inline paddle/fluid/operators/detection/multiclass_nms_op.cc]: NmsRoisNum should be always {-1} ?

Comment (yghstill) [reply paddle/fluid/operators/detection/multiclass_nms_op.cc]: Done

**Review by jerrywgz — DISMISSED** [review]
LGTM

**Review by zhiqiu — DISMISSED** [review]

**Review by jerrywgz — APPROVED** [review]
LGTM

**Review by zhiqiu — APPROVED** [review]

**Review by luotao1 — COMMENTED** [review]
@yghstill @jerrywgz 这个PR多上传了 tools/static_mode_white_list.pyc，已经在 https://github.com/PaddlePaddle/Paddle/pull/28449/files 删了

**Pull request merged**